imgproc: add image_grad using separable Sobel filters

Anishp-cell opened the pull request:
this pr adds `image_grad` to `kornia-imgproc`, computing first-order image
derivatives (dx, dy) using separable Sobel filters.
- reuses existing Sobel kernels
- computes dx and dy directly (no magnitude)
- fully documented (crate uses `deny(missing_docs)`)
- all unit tests and doc-tests pass
issue no. : #172

Anishp-cell commented:
Added a unit test for image_grad validating separable Sobel gradients on a synthetic image.
All tests and doc-tests now pass locally.
Happy to refine assertions further if needed.

edgarriba commented on crates/kornia-imgproc/src/filter/ops.rs:
```diff
@@ -586,3 +636,31 @@ mod tests {
+    image_grad(&img, &mut dx, &mut dy, 3)?;
+
+    // sanity checks (not exact equality yet)
+    assert!(dx.as_slice().iter().any(|&v| v != 0.0));
```
can you provide as expected vaues something computed with any known libraru such opencv ?

Anishp-cell replied:
Yes, sure, I’ll update the test to compare against expected Sobel gradients computed using OpenCV on the same synthetic image, with a numerical tolerance. Will push an updated test shortly.

edgarriba commented on crates/kornia-imgproc/src/filter/ops.rs:
```diff
@@ -586,3 +636,61 @@ mod tests {
+
+    let size = ImageSize {
+        width: 5,
+        height: 5,
```
Can you try with a non non squared image?

edgarriba commented on crates/kornia-imgproc/src/filter/ops.rs:
```diff
@@ -586,3 +636,61 @@ mod tests {
+    };
+
+    // 5x5 horizontal ramp image
+    let data = vec![
```
Can we include an image that produces non zeros in both dx dy ?

edgarriba commented on crates/kornia-imgproc/src/filter/ops.rs:
```diff
@@ -586,3 +636,61 @@ mod tests {
+
+    let w = size.width as usize;
+    let h = size.height as usize;
+// we compare only the interior region since border handling
```
```suggestion
  // we compare only the interior region since border handling
```
Format

Anishp-cell commented:
i updated the image_grad documentation to include a clear # Errors section for size mismatches and a # Panics section documenting the behavior for unsupported kernel sizes(3 or 5 only), keeping it consistent with existing sobel apis and avoiding introducing new error variants.
I also moved the opencv comparison test inside the #[cfg(test)] module as suggested.
All tests pass locally.

Anishp-cell commented:
All requested changes have been applied.
Fixed Clippy warnings and documentation issues
Validated Sobel kernel size to avoid panics
Ran cargo fmt, cargo clippy -D warnings, and cargo test locally
CI should now pass cleanly.

edgarriba commented on crates/kornia-imgproc/src/filter/ops.rs:
```diff
@@ -89,16 +89,66 @@ pub fn gaussian_blur<const C: usize, A1: ImageAllocator, A2: ImageAllocator>(
+/// * `dx`  - Output image storing gradient in x direction, shape (H, W, C).
+/// * `dy`  - Output image storing gradient in y direction, shape (H, W, C).
+/// * `kernel_size` - Sobel kernel size (e.g. 3 or 5).
+///
```
don't forget this

AnilKumarSingh9856 commented:
hey @edgarriba , could u please assign this issue to me? I would like to work on it.
